jscomp/others/js.ml

```diff
@@ -96,7 +96,12 @@ type +'a undefined
   A value of this type can be either undefined or 'a. This type is equivalent to Js.Undefined.t.
 *)
 
-type +'a nullable
+type +'a nullable =
+  | Present of 'a
```

Comment:
This looks great.
The only question is about naming.
Can't remember where I tool the examples from -- probably just made them up.
So if there's any thoughts about names, this is probably the right time to have them.
`Present` looks good to me, though I haven't thought deeply about alternatives.

Reply:
Maybe ask you assistant and see what comes out?

Reply:
The oracle has spoken (only keeping the suggestions that were not immediately discarded):

Here's a comparison of `Value` and `Present` with their respective advantages and disadvantages:

`Value`:
Advantages:
1. Clarity: The term "Value" directly communicates that the constructor holds a value, making it easy to understand its purpose.
2. Language-agnostic: "Value" is not tied to any specific programming language, so it doesn't carry any unintended associations.
3. Consistency: "Value" is a commonly used term in programming to denote the content of a variable or data structure, making it familiar to most developers.

Disadvantages:
1. Ambiguity: The term "Value" might be too generic, potentially causing confusion when used in different contexts or with other types.
2. Overuse: Since "Value" is a common term in programming, it might already be used elsewhere in the codebase, leading to potential conflicts or confusion.
3. Lack of distinction: "Value" does not explicitly emphasize the presence or absence of a value, unlike terms such as "Some" or "Just."

`Present`:
Advantages:
1. Specificity: The term "Present" clearly indicates that a value is present within the constructor, making its purpose explicit.
2. Uniqueness: "Present" is not a commonly used term for type constructors, making it less likely to conflict with other naming conventions in the codebase.
3. Contrast: "Present" inherently suggests a counterpart (e.g., "Absent," "Null," or "Undefined"), making it easier to understand the type's overall structure.

Disadvantages:
1. Unfamiliarity: "Present" is not a widely used term in programming languages, so developers might find it less intuitive compared to more common terms like "Value."
2. Ambiguity: Without context, the term "Present" might be less clear in its meaning than more explicit terms like "Some" or "Just."
3. Potential confusion: If other types in the codebase use similar terms (e.g., "Absent"), developers might accidentally mix up the names when working with these types.

Both `Value` and `Present` have their pros and cons. The choice between them will depend on the specific context and the conventions and style of the ReScript language and its codebase. Consider the trade-offs and choose the option that best fits the overall design and goals of the project.

Reply:
A bit more far fetched:

In the context of ReScript and JavaScript, Promises are used to represent asynchronous computations that might eventually return a value or an error. While nullable types and Promises are different concepts, it's possible to find similarities in how they handle values and their absence.

For instance, both nullable types and Promises deal with values that might or might not be available:

1. Nullable types: They represent values that might be `null` or `undefined`. The use of constructors like `Present` (or `Value`) can denote the presence of a value, whereas other constructors like `Null` or `Undefined` indicate its absence.
2. Promises: They represent asynchronous computations that might eventually return a value or an error. When a Promise resolves, it contains a value, and when it rejects, it contains an error.

If you'd like to unify the concepts for a more consistent naming convention, you can consider using similar names for the constructors in nullable types and the resolved or rejected states of Promises:

1. For the presence of a value, you can use `Value`, `Present`, `Resolved`, or `Fulfilled`. This would emphasize the successful completion of a Promise or the availability of a value in a nullable type.
2. For the absence of a value or presence of an error, you can use `Error`, `Rejected`, or `Failed`. This would underline that the Promise has failed or that the nullable type does not contain a value.

While these similarities can be helpful, keep in mind that nullable types and Promises are different concepts with different use cases. Unifying naming conventions can improve clarity, but it's essential to ensure that the distinctions between the concepts are not lost or obscured.

Reply:
Value it is.